File: apps/ios/dakka/Core/Crypto/E2E.swift
import Foundation
import CryptoKit

/// E2E-шифрование, бит-в-бит совместимое с веб-клиентом
/// (`packages/web/src/lib/crypto.ts` + `e2e.ts`).
///
/// Алгоритм:
/// - **Обмен ключами**: ECDH P-256 (NIST). Веб использует WebCrypto
///   `ECDH P-256`, у нас — CryptoKit `P256.KeyAgreement`.
/// - **Симметричное шифрование**: AES-256-GCM. Веб использует
///   `AES-GCM 256bit`, у нас — `CryptoKit.AES.GCM`.
/// - **Дериваця AES-ключа**: WebCrypto при `deriveKey(ECDH → AES-GCM)`
///   берёт **X-координату ECDH-секрета** напрямую как AES-ключ
///   (без HKDF). У `P256.KeyAgreement.SharedSecret` это ровно 32 байта —
///   они и есть AES-ключ.
/// - **Кодирование ключей**: SPKI для публичного, PKCS#8 для приватного,
///   обёрнутые в base64. CryptoKit отдаёт `derRepresentation` именно в этих
///   форматах — interop гарантирован.
enum E2E {

    // MARK: - Errors

    enum CryptoError: LocalizedError {
        case invalidPrivateKey
        case invalidPublicKey
        case invalidCiphertext
        case decryptionFailed

        var errorDescription: String? {
            switch self {
            case .invalidPrivateKey:  return "Некорректный приватный ключ"
            case .invalidPublicKey:   return "Некорректный публичный ключ"
            case .invalidCiphertext:  return "Некорректный шифртекст"
            case .decryptionFailed:   return "Не удалось расшифровать сообщение"
            }
        }
    }

    // MARK: - Key generation

    struct KeyPair {
        let publicKeyB64:  String
        let privateKeyB64: String
    }

    static func generateKeyPair() -> KeyPair {
        let priv = P256.KeyAgreement.PrivateKey()
        return KeyPair(
            publicKeyB64:  priv.publicKey.derRepresentation.base64EncodedString(),
            privateKeyB64: priv.derRepresentation.base64EncodedString()
        )
    }

    // MARK: - Encryption

    struct Encrypted {
        let ciphertextB64: String
        let nonceB64:      String
    }

    /// Шифрует UTF-8 текст. Возвращает (ciphertext, nonce) — оба base64.
    /// `senderPrivateKeyB64`, `recipientPublicKeyB64` — те же base64-форматы,
    /// что отдаёт `generateKeyPair`.
    static func encryptText(
        _ plaintext: String,
        recipientPublicKeyB64: String,
        senderPrivateKeyB64: String
    ) throws -> Encrypted {
        let sharedKey = try deriveSharedKey(
            theirPublicKeyB64: recipientPublicKeyB64,
            myPrivateKeyB64:   senderPrivateKeyB64
        )

        // WebCrypto AES-GCM nonce — 96 бит / 12 байт
        let nonce = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(
            Data(plaintext.utf8),
            using: sharedKey,
            nonce: nonce
        )
        // sealed.ciphertext — голый шифртекст БЕЗ tag,
        // sealed.tag — 16 байт authenticator.
        // WebCrypto AES-GCM по умолчанию конкатенирует ciphertext+tag,
        // поэтому склеиваем тоже.
        let combined = sealed.ciphertext + sealed.tag
        return Encrypted(
            ciphertextB64: combined.base64EncodedString(),
            nonceB64:      Data(nonce).base64EncodedString()
        )
    }

    /// Расшифровывает. Возвращает текст или бросает.
    static func decryptText(
        ciphertextB64: String,
        nonceB64: String,
        senderPublicKeyB64: String,
        recipientPrivateKeyB64: String
    ) throws -> String {
        guard let cipher = Data(base64Encoded: ciphertextB64),
              let nonceData = Data(base64Encoded: nonceB64)
        else { throw CryptoError.invalidCiphertext }

        // WebCrypto-encoded: последние 16 байт — это GCM-tag.
        guard cipher.count > 16 else { throw CryptoError.invalidCiphertext }
        let tag = cipher.suffix(16)
        let ct  = cipher.prefix(cipher.count - 16)

        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sharedKey = try deriveSharedKey(
            theirPublicKeyB64: senderPublicKeyB64,
            myPrivateKeyB64:   recipientPrivateKeyB64
        )

        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ct, tag: tag)
        let plain: Data
        do {
            plain = try AES.GCM.open(box, using: sharedKey)
        } catch {
            throw CryptoError.decryptionFailed
        }

        guard let str = String(data: plain, encoding: .utf8) else {
            throw CryptoError.decryptionFailed
        }
        return str
    }

    // MARK: - Internal: shared key derivation + cache

    /// Кэш выведенных AES-ключей — экономит ECDH-операции при потоке сообщений.
    /// Ключ кэша: первые 16 символов `theirPublicKeyB64`.
    private static var sharedKeyCache: [String: SymmetricKey] = [:]
    private static let cacheLock = NSLock()

    private static func deriveSharedKey(
        theirPublicKeyB64: String,
        myPrivateKeyB64: String
    ) throws -> SymmetricKey {
        let cacheKey = String(theirPublicKeyB64.prefix(16))
        cacheLock.lock(); defer { cacheLock.unlock() }
        if let cached = sharedKeyCache[cacheKey] { return cached }

        guard let theirPubData = Data(base64Encoded: theirPublicKeyB64)
        else { throw CryptoError.invalidPublicKey }
        guard let myPrivData = Data(base64Encoded: myPrivateKeyB64)
        else { throw CryptoError.invalidPrivateKey }

        let theirPub: P256.KeyAgreement.PublicKey
        do {
            theirPub = try P256.KeyAgreement.PublicKey(derRepresentation: theirPubData)
        } catch { throw CryptoError.invalidPublicKey }

        let myPriv: P256.KeyAgreement.PrivateKey
        do {
            myPriv = try P256.KeyAgreement.PrivateKey(derRepresentation: myPrivData)
        } catch { throw CryptoError.invalidPrivateKey }

        let shared = try myPriv.sharedSecretFromKeyAgreement(with: theirPub)

        // X-координата (32 байта для P-256) — напрямую как AES-256 key.
        // SharedSecret конформит ContiguousBytes.
        let key: SymmetricKey = shared.withUnsafeBytes { rawBuf in
            SymmetricKey(data: Data(rawBuf))
        }

        sharedKeyCache[cacheKey] = key
        return key
    }

    /// Полная очистка кэша — вызывать при logout.
    static func clearCache() {
        cacheLock.lock(); defer { cacheLock.unlock() }
        sharedKeyCache.removeAll()
    }
}

File: apps/android/app/src/main/java/online/akkdmsg/dakka/data/crypto/E2E.kt
package online.akkdmsg.dakka.data.crypto

import android.util.Base64
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.SecureRandom
import java.security.spec.ECGenParameterSpec
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * E2E-шифрование Android-клиента — байт-в-байт совместимое с iOS (CryptoKit)
 * и web (WebCrypto).
 *
 * Алгоритм:
 *   • ECDH P-256 (secp256r1) для обмена ключами
 *   • AES-256-GCM для симметричного шифрования
 *   • Raw X-coordinate (32 байта) ECDH-секрета используется как AES-key
 *     БЕЗ HKDF — потому что так делает WebCrypto AES-GCM deriveKey.
 *
 * Форматы ключей (для interop):
 *   • Public:  SPKI (X.509 encoded), base64
 *   • Private: PKCS#8, base64
 *   Это нативные форматы Android JCA + WebCrypto + CryptoKit derRepresentation.
 *
 * AES-GCM:
 *   • Nonce 12 байт (96 бит) — стандарт для GCM
 *   • Output WebCrypto: ciphertext || tag (где tag 16 байт) — JCA по умолчанию
 *     возвращает ровно это же.
 */
object E2E {

    sealed class CryptoError(msg: String) : Exception(msg) {
        object InvalidPrivateKey : CryptoError("Некорректный приватный ключ")
        object InvalidPublicKey  : CryptoError("Некорректный публичный ключ")
        object InvalidCiphertext : CryptoError("Некорректный шифртекст")
        object DecryptionFailed  : CryptoError("Не удалось расшифровать сообщение")
    }

    data class KeyPair(val publicKeyB64: String, val privateKeyB64: String)
    data class Encrypted(val ciphertextB64: String, val nonceB64: String)

    private val keyFactory: KeyFactory = KeyFactory.getInstance("EC")

    private val sharedKeyCache: MutableMap<String, SecretKeySpec> = mutableMapOf()
    private val cacheLock = Any()
    private val secureRandom = SecureRandom()

    // MARK: - Key generation

    fun generateKeyPair(): KeyPair {
        val kpg = KeyPairGenerator.getInstance("EC")
        kpg.initialize(ECGenParameterSpec("secp256r1"))
        val pair = kpg.generateKeyPair()
        return KeyPair(
            publicKeyB64  = Base64.encodeToString(pair.public.encoded,  Base64.NO_WRAP),
            privateKeyB64 = Base64.encodeToString(pair.private.encoded, Base64.NO_WRAP),
        )
    }

    // MARK: - Encrypt

    @Throws(CryptoError::class)
    fun encryptText(
        plaintext: String,
        recipientPublicKeyB64: String,
        senderPrivateKeyB64: String,
    ): Encrypted {
        val sharedKey = deriveSharedKey(recipientPublicKeyB64, senderPrivateKeyB64)
        val nonce = ByteArray(12).also(secureRandom::nextBytes)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, sharedKey, GCMParameterSpec(128, nonce))
        val combined = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))
        // JCA AES/GCM/NoPadding по умолчанию даёт ciphertext || 16-byte tag — то же
        // что WebCrypto. Совместимо.

        return Encrypted(
            ciphertextB64 = Base64.encodeToString(combined, Base64.NO_WRAP),
            nonceB64      = Base64.encodeToString(nonce,    Base64.NO_WRAP),
        )
    }

    // MARK: - Decrypt

    @Throws(CryptoError::class)
    fun decryptText(
        ciphertextB64: String,
        nonceB64: String,
        senderPublicKeyB64: String,
        recipientPrivateKeyB64: String,
    ): String {
        val ciphertext = runCatching { Base64.decode(ciphertextB64, Base64.NO_WRAP) }
            .getOrElse { throw CryptoError.InvalidCiphertext }
        val nonce = runCatching { Base64.decode(nonceB64, Base64.NO_WRAP) }
            .getOrElse { throw CryptoError.InvalidCiphertext }
        if (ciphertext.size <= 16) throw CryptoError.InvalidCiphertext

        val sharedKey = deriveSharedKey(senderPublicKeyB64, recipientPrivateKeyB64)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, sharedKey, GCMParameterSpec(128, nonce))

        val plain = try {
            cipher.doFinal(ciphertext)
        } catch (_: Exception) {
            throw CryptoError.DecryptionFailed
        }
        return String(plain, Charsets.UTF_8)
    }

    // MARK: - Shared key (ECDH P-256) с кэшем

    @Throws(CryptoError::class)
    private fun deriveSharedKey(
        theirPublicKeyB64: String,
        myPrivateKeyB64: String,
    ): SecretKeySpec {
        // CRITICAL: cache key должен зависеть ОТ ВСЕГО публичного ключа.
        // У всех P-256 SPKI base64 первые ~36 символов идентичны (DER-префикс),
        // поэтому take(16) возвращал бы одинаковую строку для разных пиров
        // → cross-chat confidentiality break. Используем SHA-256(full pub).
        val cacheKey = sha256Hex(theirPublicKeyB64)
        synchronized(cacheLock) { sharedKeyCache[cacheKey] }?.let { return it }

        val theirPubBytes = runCatching { Base64.decode(theirPublicKeyB64, Base64.NO_WRAP) }
            .getOrElse { throw CryptoError.InvalidPublicKey }
        val myPrivBytes = runCatching { Base64.decode(myPrivateKeyB64, Base64.NO_WRAP) }
            .getOrElse { throw CryptoError.InvalidPrivateKey }

        val theirPub = runCatching {
            keyFactory.generatePublic(X509EncodedKeySpec(theirPubBytes))
        }.getOrElse { throw CryptoError.InvalidPublicKey }

        val myPriv = runCatching {
            keyFactory.generatePrivate(PKCS8EncodedKeySpec(myPrivBytes))
        }.getOrElse { throw CryptoError.InvalidPrivateKey }

        val agreement = KeyAgreement.getInstance("ECDH")
        agreement.init(myPriv)
        agreement.doPhase(theirPub, true)
        val sharedSecret = agreement.generateSecret()  // 32 байта X-coord для P-256

        val key = SecretKeySpec(sharedSecret, "AES")
        synchronized(cacheLock) { sharedKeyCache[cacheKey] = key }
        return key
    }

    /** Очистить кэш — вызывать при logout. */
    fun clearCache() {
        synchronized(cacheLock) { sharedKeyCache.clear() }
    }

    private fun sha256Hex(s: String): String {
        val md = java.security.MessageDigest.getInstance("SHA-256")
        val digest = md.digest(s.toByteArray(Charsets.UTF_8))
        val sb = StringBuilder(digest.size * 2)
        for (b in digest) sb.append("%02x".format(b))
        return sb.toString()
    }
}
